packages/flutter/lib/src/cupertino/dialog.dart

```diff
@@ -410,106 +575,686 @@ class _CupertinoAlertActionSection extends StatelessWidget {
+// behavior for its buttons. It only handles the sizing and layout of buttons.
+// Scrolling behavior can be composed on top of this widget, if desired.
 //
-// Draws the cross-axis divider lines, used when the layout is horizontal.
-class _CupertinoHorizontalDividerPainter extends CustomPainter {
-  _CupertinoHorizontalDividerPainter(this.count);
+// [_CupertinoDialogActionsRenderWidget] requires an ancestor of type
+// [_DialogInteractionInheritedWidget] so that the rendering of dividers can
+// depend on a user's interaction.
+//
+// See [_RenderCupertinoDialogActions] for specific layout policy details.
+class _CupertinoDialogActionsRenderWidget extends MultiChildRenderObjectWidget {
+  _CupertinoDialogActionsRenderWidget({
+    Key key,
+    @required List<Widget> actionButtons,
+    @required Set<int> pressedButtons,
+    double dividerWidth = 0.0,
+  }) : assert(pressedButtons != null),
+        _pressedButtons = pressedButtons,
+        _dividerWidth = dividerWidth,
+        super(key: key, children: actionButtons);
 
-  final int count;
+  final Set<int> _pressedButtons;
+  final double _dividerWidth;
 
   @override
-  void paint(Canvas canvas, Size size) {
-    final Paint paint = new Paint()..color = _kButtonDividerColor;
-
-    canvas.drawLine(Offset.zero, new Offset(size.width, 0.0), paint);
-    for (int i = 1; i < count; ++i) {
-      // TODO(abarth): Hide the divider when one of the adjacent buttons is
-      // highlighted.
-      final double x = size.width * i / count;
-      canvas.drawLine(new Offset(x, 0.0), new Offset(x, size.height), paint);
-    }
+  RenderObject createRenderObject(BuildContext context) {
+    return new RenderCupertinoDialogActions(
+      pressedButtons: _pressedButtons,
+      dividerWidth: _dividerWidth,
+    );
   }
 
   @override
-  bool shouldRepaint(_CupertinoHorizontalDividerPainter other) => count != other.count;
+  void updateRenderObject(BuildContext context, RenderCupertinoDialogActions renderObject) {
+    renderObject.pressedButtons = _pressedButtons;
+    renderObject.dividerWidth = _dividerWidth;
+  }
 }
 
-// A CustomPainter to draw the divider lines.
-//
-// Draws the cross-axis divider lines, used when the layout is vertical.
-class _CupertinoVerticalDividerPainter extends CustomPainter {
-  _CupertinoVerticalDividerPainter();
+/// iOS style layout policy for sizing and positioning an alert dialog's action
+/// buttons.
+///
+/// The policy is as follows:
+///
+/// If a single action button is provided, or if 2 action buttons are provided
+/// that can fit side-by-side, then action buttons are sized and laid out in a
+/// single horizontal row. The row is exactly as wide as the dialog, and the row
+/// is as tall as the tallest action button. A horizontal divider is drawn above
+/// the button row. If 2 action buttons are provided, a vertical divider is
+/// drawn between them. The thickness of the divider is set by [dividerWidth].
```

Comment:
nit: maybe just call this dividerThickness if it's used both horizontally and vertically.

Reply:
done